### PR TITLE
Generate properties for optional query parameters

### LINKED ISSUE #67: Pubsub Schema creation fails with "Invalid resource name given"
When attempting to create a pubsub schema with python, I'm getting an error "Invalid resource name given (name=projects/my-gcp-project/schemas/)" even though I'm specifying the resource_name and name params with the pattern: projects/my-gcp-project/schemas/someuniquename

## Expected behavior
The schema should be created without error.

## Current behavior
I'm getting the following error: google-native:pubsub/v1:Schema (someuniquename):
    error: error sending request: googleapi: Error 400: Invalid resource name given (name=projects/my-gcp-project/schemas/). Refer to https://cloud.google.com/pubsub/docs/admin#resource_names for more information.: 

## Steps to reproduce

1. Add this code: schema = pubsub_native.v1.Schema(resource_name=schemas_id, projects_id=GCP_PROJECT_ID,
                                 schemas_id=schemas_id, definition=json.dumps(sch1), name=schema_name, type='avro')
2. pulumi up
3. yes

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Generate properties for optional query parameters instead of ignoring them. Also, extend the metadata to keep the location of the parameters and format the URL accordingly.

Resolves #67 - the new `schemaId` property is actually required there, even though it's not marked as such in the DD.

### CODE REVIEW

File: provider/pkg/provider/provider.go [hunk 1]
```diff
@@ -220,7 +221,7 @@ func (p *googleCloudProvider) Diff(_ context.Context, req *rpc.DiffRequest) (*rp
 		}
 	}
 
-	return &rpc.DiffResponse{Changes: rpc.DiffResponse_DIFF_UNKNOWN, Replaces: replaces}, nil
+	return &rpc.DiffResponse{Changes: rpc.DiffResponse_DIFF_UNKNOWN, Replaces: replaces, DeleteBeforeReplace: true}, nil
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:224]: Why is this change needed? Isn't this a major behavior change?

File: provider/pkg/provider/provider.go [hunk 2]
```diff
@@ -291,10 +292,30 @@ func (p *googleCloudProvider) Create(ctx context.Context, req *rpc.CreateRequest
...
+				continue
+			}
+
 			value := inputs[key].StringValue()
-			path = strings.Replace(path, fmt.Sprintf("{%s}", param), value, 1)
+			switch param.Location {
+			case "path":
+				path = strings.Replace(path, fmt.Sprintf("{%s}", param.Name), url.PathEscape(value), 1)
+			case "query":
+				path = fmt.Sprintf("%s%s%s=%s", path, querySep, key, url.QueryEscape(value))
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:314]: Any reason not to use something like [url.Url](https://golang.org/pkg/net/url/#URL) to construct the URL?

File: provider/pkg/provider/provider.go (hunk 1 above)
Comment (mikhailshilkov) [reply provider/pkg/provider/provider.go:224]: Ah, I should have commented on that. It's unrelated to this PR, I just noticed that replaces are failing while I was testing. We don't have auto-naming yet, so any resource needs to be deleted first to free up its name.

File: provider/pkg/provider/provider.go (hunk 2 above)
Comment (mikhailshilkov) [reply provider/pkg/provider/provider.go:314]: Not really. I wasn't sure how to combine it with replace operations on the path above.

Comment (mikhailshilkov) [reply provider/pkg/provider/provider.go]: Good point! I refactored the code to use `url` to build the query.

**Review by lblackstone — APPROVED** [review]

**Pull request merged**